### PR TITLE
MM-48089 Fix broken links in emails

### PR DESCRIPTION
**Pull request by m-ripper**
#### Summary

Transform relative links into absolute links when converting Markdown to HTML.
This will fix invalid links in emails.

#### Ticket Link

[MM-48089](https://mattermost.atlassian.net/browse/MM-48089)

#### Release Note

```release-note
Fixed broken links in emails
```

[MM-48089]: https://mattermost.atlassian.net/browse/MM-48089?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: utils/markdown.go [hunk 1]
```diff
@@ -36,10 +36,17 @@ func StripMarkdown(markdown string) (string, error) {
 }
 
 // MarkdownToHTML takes a string containing Markdown and returns a string with HTML tagged version
-func MarkdownToHTML(markdown string) (string, error) {
+func MarkdownToHTML(markdown, siteURL string) (string, error) {
+	// Turn relative links into absolute links
+	relLinkRe := regexp.MustCompile(`\[(.*)]\((/.*)\)`)
```
Comment (agnivade) [inline utils/markdown.go:41]: Let's pre-compile this regex outside. Compiling regexes every time has a performance cost: https://developers.mattermost.com/contribute/more-info/server/style-guide/#performance-sensitive-areas

File: utils/markdown.go [hunk 2]
```diff
@@ -36,10 +36,17 @@ func StripMarkdown(markdown string) (string, error) {
...
+	absLinkMarkdown := relLinkRe.ReplaceAllFunc([]byte(markdown), func(s []byte) []byte {
+		out := relLinkRe.ReplaceAllString(string(s), "[$1]("+siteURL+"$2)")
+		return []byte(out)
+	})
+
 	// Unescape any blockquote text to be parsed by the markdown parser.
 	re := regexp.MustCompile(`^|\n(&gt;)`)
```
Comment (agnivade) [inline utils/markdown.go:48]: Actually, this as well while we are here :)

File: utils/markdown.go [hunk 3]
```diff
@@ -35,11 +35,19 @@ func StripMarkdown(markdown string) (string, error) {
...
 }
 
+var relLinkReg = regexp.MustCompile(`\[(.*)]\((/.*)\)`)
+var blockquoteReg = regexp.MustCompile(`^|\n(&gt;)`)
+
 // MarkdownToHTML takes a string containing Markdown and returns a string with HTML tagged version
-func MarkdownToHTML(markdown string) (string, error) {
+func MarkdownToHTML(markdown, siteURL string) (string, error) {
+	// Turn relative links into absolute links
+	absLinkMarkdown := relLinkReg.ReplaceAllFunc([]byte(markdown), func(s []byte) []byte {
```
Comment (agnivade) [inline utils/markdown.go:44]: Let's use https://pkg.go.dev/regexp#Regexp.ReplaceAllStringFunc to avoid the double conversion.

**Review by agnivade — APPROVED** [review]
Thanks!

**Comment by amyblais** [comment]
@m-ripper This could optionally be cherry-picked to the ``cloud`` so it will go to the next cloud and self-hosted releases. Please open a manual cherry-pick PR against the ``cloud`` branch if you have time.

**Pull request merged**